src/hooks/useMoveThoughtAnimation.ts

```diff
@@ -0,0 +1,132 @@
+import { useLayoutEffect, useMemo, useState } from 'react'
+import { shallowEqual, useSelector } from 'react-redux'
+import State from '../@types/State'
+import durations from '../durations.config'
+import attributeEquals from '../selectors/attributeEquals'
+import usePrevious from './usePrevious'
+
+/**
+ * Return type of useMoveThoughtAnimation.
+ */
+export interface MoveThoughtAnimation {
+  /** The style to apply to the move div. */
+  moveDivStyle: React.CSSProperties | undefined
+}
+
+interface Options {
+  /** The current on-screen index of the thought. */
+  index: number
+}
+
+/**
+ * Encapsulates the logic that determines the animation state when a thought is moved
+ * up or down via the moveThoughtUp / moveThoughtDown reducers. This hook abstracts all
+ * of the Redux selectors and timing needed so that the consuming component (TreeNode)
+ * only has to deal with the resulting animation flags.
+ */
+const useMoveThoughtAnimation = ({ index }: Options): MoveThoughtAnimation => {
+  const { lastMoveType, isTableView, isCrossContext } = useSelector((state: State) => {
+    const lastPatches = state.undoPatches[state.undoPatches.length - 1]
+
+    const lastMoveType = lastPatches?.some(patch => ['moveThoughtUp', 'moveThoughtDown'].includes(patch.actions[0]))
+      ? (lastPatches[0].actions[0] as 'moveThoughtUp' | 'moveThoughtDown')
+      : null
+
+    const cursor = state.cursor
+    const isTableView = cursor ? attributeEquals(state, cursor[0], '=view', 'Table') : false
+    const isCrossContext = (lastPatches ?? []).some(patch => patch.path?.endsWith('/parentId'))
+
+    return {
+      lastMoveType,
+      isTableView,
+      isCrossContext,
+    }
```

Comment:
By returning a composite object, you're missing the value of combining multiple `useSelector` into a single `useSelector`. Composite object return types are ugly and should be avoided anyway. There's no elegance in a composite object of arbitrary values.

Notice how `isTableView` and `isCrossContext` are *only ever used in conjunction* in the expression `isCrossContext && isTableView`. You do not need to return them independently. It's less encapsulated, and it's worse performance because the memo will be refreshed when *either* of them change.

*Encapsulate* to hide complexity. *Combine* logic early to improve performance and reduce complexity downstream.

Reply:
I’ve updated the hook to return a tuple `[lastMoveType, skipMoveAnimation]` instead of a composite object. I combined `isTableView` and `isCrossContext` early into skipMoveAnimation to keep things cleaner and avoid extra memo updates. Let me know if anything still feels off.

https://github.com/cybersemics/em/blob/b497f9b43f2e99639e1f6d0c4da2e80d2c42aaa0/src/hooks/useMoveThoughtAnimation.ts#L26-L44

Reply:
Okay, thanks.

FYI A tuple is a composite object, but worse because the entries are unnamed.